Initial ES6 port

ibgreen opened the pull request:
Examples run. Initial build scripts in place. 

Keeping on es6 branch since additional testing and script polish needed.

ibgreen commented:
> there may be a few other places were `opts` is used. We could replace these with spread syntax. ie., `({fill, stroke})`.
Done

vicapow commented:
@ibgreen looking good so far. What do you think about adding `dist/` to `.gitignore` ? I think it would cut down on the diff noise here and in the future.

Qix- commented:
:+1: on ignoring `dist/` - they'll cause desynchronization and noisy diffs otherwise since they're generated.

ibgreen commented:
> What do you think about adding dist/ to .gitignore
@vicapow @Qix - Yes good idea. 
- added `dist` to `.gitgnore`
-  added `.npmignore` file to override `.gitignore` for npm package publishing (which uses gitignore by default when selecting which files to publish). 
- Can't be completely sure until we publish a version but it should work.

All examples work per my testing. 
I cannot get unit tests to work (npm issues?), but pushed a new commit anyway per discussion with @vicapow.

vicapow commented on example/examples/custom.react.js:
```diff
@@ -66,85 +69,97 @@ var OverlayExample = React.createClass({
+          redraw={ this._redrawCanvasOverlay }/>
+
+        {
+          /* We use invisible SVG elements to support interactivity. */
```
can you move this comment back to the relevant code?

vicapow commented on example/examples/route.react.js:
```diff
@@ -53,70 +56,77 @@ var RouteOverlayExample = React.createClass({
+  }
+
+  @autobind
+  _redrawCanvasOverlay(opt) {
```
good use of `({ctx, width, height})` instead of `(opt)`

vicapow commented on example/examples/style-diffing.react.js:
```diff
@@ -17,61 +17,61 @@
-    width: React.PropTypes.number.isRequired,
-    height: React.PropTypes.number.isRequired
-  },
+  static get propTypes() {
```
Is there any benefit to `static get propTypes() {` over...

````js
StyleDiffingExample.propTypes = {...};
````
If not, the latter seems more inline with the React docs.

ibgreen replied:
> Is there any benefit to `static get propTypes() {` over...`StyleDiffingExample.propTypes = {...};`
I feel it is important for readability that the public api (i.e. the props) comes first, just like in an ES5 react component. Whether to compensate for the lack of static properties in ES6 with a static getter at the beginning of the class, or an assignment after the end is a matter of taste. I dropped the static accessors per your suggestion, but left the PROP_TYPES constants at the top, and did a `StyleDiffingExample.propTypes = PROP_TYPES;` at the bottom. Does that seem reasonable?

chrisirhc replied:
👍  to `PROP_TYPES`

vicapow commented on src/config.js:
```diff
@@ -17,22 +17,19 @@
   CURSOR: {
-    GRABBING: prefix + 'grabbing',
-    GRAB: prefix + 'grab'
+    GRABBING: `${PREFIX} grabbing`,
```
You've got an extra space here. should be `${PREFIX}grabbing`. ditto for the next line. otherwise, the "grabber" cursor wont be present.

vicapow commented on src/map-interactions.react.js:
```diff
@@ -187,36 +195,39 @@ var MapInteractions = React.createClass({
-    this.props.onZoom({pos: pos, scale: scale});
+    this.props.onZoom({pos, scale});
     window.clearTimeout(this._zoomEndTimeout);
     this._zoomEndTimeout = window.setTimeout(function _setTimeout() {
```
the closure can be `() => this.props.onZoomEnd()`

ibgreen replied:
> the closure can be () => this.props.onZoomEnd()

I've had problems when using window.setTimeout with fat arrow functions in the past, and since they are asynchronous they can be a little tricky to test. So I intentionally left these out for now. Easy enough to update them when everything else is confirmed to be working.

vicapow commented on src/map.react.js:
```diff
@@ -531,22 +506,22 @@ var MapGL = React.createClass({
-    var map = this._getMap();
-    var transform = cloneTransform(map.transform);
-    var around = unprojectFromTransform(transform, opt.pos);
+  @autobind _onZoom(opt) {
```
`_onZoom({scale, pos})`

vicapow commented on src/map.react.js:
```diff
@@ -555,81 +530,112 @@ var MapGL = React.createClass({
-    var style = assign({}, props.style, {
+  render() {
+    const {className} = this.props;
+    const props = this.props;
```
````js
const {props} = this;
const {className} = props;
````

vicapow commented on test/index.js:
```diff
@@ -1,11 +1,12 @@
-'use strict';
+require('babel/register');
```
I think this is missing from `package.json`

vicapow commented:
@ibgreen almost there! I left some comments inline. Can you also squash these commits into one to clear the history with the `dist` commits?

ibgreen commented:
> almost there! I left some comments inline. Can you also squash these commits into one to clear the history with the dist commits?
@vicapow - Squashed to single commit, dist removed. I believe that I have either resolved or responded to all your comments, let me know if I missed something. If not, it might be good time to pull this into your es6 branch so that you can do some testing on your side, and I can open new PRs as needed.

vicapow commented:
@ibgreen I was about to merge but then noticed some errors while testing. Specifically, I don't think this will work for non-babel consumers. I'd like to still support ES5 consumers but if others disagree, I'm happy to move forward with this.

Specifically, looking at dist/index.js, I see...

````js
Object.defineProperty(exports, "__esModule", {
  value: true
});

var _map = require('./map.react');

Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_map).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
````

Which for commonJS / es5 consumers exposes the following object when requiring.

````js
var MapGL = require('../');
console.log(MapGL); // { __esModule: true, get default: function() { ... }, __proto__: ...}
````

This is causing the tests to fail for me. I know I originally said to forget the tests, but now I think it's important to make sure the tests are running for you before this lands, to make sure the `require` in ES5 works properly. We can sync up offline.

> If not, it might be good time to pull this into your es6 branch so that you can do some testing on your side, and I can open new PRs as needed.

Sure. Feel free to merge into the es6 branch for now. If you do, can you create a new issue for the above?

ibgreen commented:
@vicapow - Top level file uses now module.exports. I will merge this into main repo.
